[Tin Ruiqi] iP

ruiqi7 opened the pull request:
### DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/ruiqi7/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

ruiqi7 committed:
Greet, echo, exit

ruiqi7 committed:
Add, list

ruiqi7 committed:
Add Task.java

ruiqi7 committed:
Add TaskList.java

ruiqi7 committed:
Mark as done / not done

ruiqi7 committed:
Add ToDo.java

ruiqi7 committed:
Add Deadline.java

ruiqi7 committed:
Add Event.java

ruiqi7 committed:
toString() for Task objects

ruiqi7 committed:
Todos, events, deadlines

ruiqi7 committed:
Automated text ui testing

ruiqi7 committed:
Add DukeException.java

ruiqi7 committed:
Add InvalidTaskException.java, MissingInfoException.java, InvalidInputException.java

ruiqi7 committed:
Handle errors

ruiqi7 committed:
Delete

ruiqi7 committed:
Add InvalidTaskNumberException.java

ruiqi7 committed:
Add methods and remove some custom exceptions from Duke.java

ruiqi7 committed:
Remove TaskList.java and some custom Exception classes

ruiqi7 committed:
Update input.txt and EXPECTED.TXT

ruiqi7 committed:
Use enums

ruiqi7 committed:
Shift exceptions into methods

ruiqi7 committed:
Use enums for all commands

ruiqi7 committed:
Add TaskList.java

ruiqi7 committed:
Add TaskFile.java for file creation, saving and reading

ruiqi7 committed:
Read in dates for deadlines and events

ruiqi7 committed:
Update TaskFile.java to create a file before reading it

ruiqi7 committed:
Merge branch 'branch-Level-7'

ruiqi7 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/TaskList.java

ruiqi7 committed:
Extract Ui, Storage, Parser and *Command classes

ruiqi7 committed:
Organize into packages

ruiqi7 committed:
Merge remote-tracking branch 'gradle/add-gradle-support'

ruiqi7 committed:
Add JUnit tests for Todo, Deadline and Event classes

ruiqi7 committed:
Create a JAR file

ruiqi7 committed:
Fixed some UI messages

ruiqi7 committed:
Add JavaDoc comments

ruiqi7 committed:
Modify code to comply with coding standard

ruiqi7 committed:
Add FindCommand.java

ruiqi7 committed:
Merge branch 'branch-A-JavaDoc'

ruiqi7 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/commands/DeadlineCommand.java
#	src/main/java/duke/commands/EventCommand.java
#	src/main/java/duke/commands/TodoCommand.java
#	src/main/java/duke/data/TaskList.java
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/tasks/Deadline.java
#	src/main/java/duke/tasks/Event.java
#	src/main/java/duke/ui/Ui.java

ruiqi7 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/common/Message.java
#	src/main/java/duke/data/TaskList.java
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/tasks/Task.java
#	src/main/java/duke/ui/Ui.java

ruiqi7 committed:
Remove redundant 'this' keyword

ruiqi7 committed:
Add GUI

ruiqi7 committed:
Merge branch 'branch-Level-10'

ruiqi7 committed:
Use CheckStyle to fix coding standard violations

ruiqi7 committed:
Merge branch 'branch-A-CheckStyle'

eugenetayyj commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,58 @@
+ * Represents the Duke chatbot.
+ */
+public class Duke {
+    private Storage storage;
```
I like how you abstracted storage into a separate storage class.

eugenetayyj commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,58 @@
+        ui = new Ui();
+        storage = new Storage();
+        try {
+            tasks = new TaskList(storage.load());
```
I like how you used camelCase naming convention here.

eugenetayyj commented on src/main/java/duke/commands/ByeCommand.java:
```diff
@@ -0,0 +1,32 @@
+package duke.commands;
+
+import duke.data.TaskList;
```
I like how the import statements are explicitly specified rather than a wildcard import statement.

eugenetayyj commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,146 @@
+public class Parser {
+    private static final String DATE_INPUT_FORMAT = "yyyy-MM-dd";
+
+    private enum MissingDetails {
```
I like the use of enum for standard inputs.

eugenetayyj reviewed (commented):
Overall, I like how you follow the coding standards in place in CS2103T!👍🏽

ruiqi7 committed:
Add missing @throws tag for some TaskList methods

ruiqi7 committed:
Use varargs to take in multiple keywords for FindCommand

ruiqi7 committed:
Merge branch 'branch-A-Varargs'

myangat0343 commented on src/main/java/duke/common/Messages.java:
```diff
@@ -0,0 +1,31 @@
+    public static final String MESSAGE_NO_SUCH_TASK = "No such task!";
+    public static final String MESSAGE_PROVIDE_TASK_NUMBER = "Please provide a task number!";
+    public static final String MESSAGE_PROVIDE_DESCRIPTION = "Please provide a task description!";
+    public static final String MESSAGE_PROVIDE_DESCRIPTION_AND_DATE = "Please provide a task description and a date!";
```
Your code is well-organized !!
Just a small suggestion here:  "Please provide the description and date of your task!"; Maybe you could adjust the sentences : )

myangat0343 commented on src/main/java/duke/data/exception/DukeException.java:
```diff
@@ -0,0 +1,11 @@
+ * Represents an exception to be thrown when invalid inputs are provided by the user
+ * or when there is an error loading or saving the list of tasks.
+ */
+public class DukeException extends Exception {
```
Hiii, I think it is good to use DukeException to handle the exception in your code. And there is a small suggestion that maybe you could use some child-classes of DukeException to handle different exceptions, like TaskNotExistException. In this way, it could be more well-organized : )

myangat0343 reviewed (commented):
Some small suggestions are given. Like your code : )
]

ruiqi7 committed:
Add assertions to Parser and Storage classes

Parser handles user input and Storage handles file input, both of
which could possibly be invalid.

Using assertions help to verify that the inputs are of sufficient
length and can be accessed safely via array indexing.

Assertions have a low impact on performance and help to identify
bugs in the code.

ruiqi7 committed:
Improve code quality

The DukeException class represents all exceptions.

Categorizing all exceptions under DukeException is not specific
enough to give a clear idea of what an exception is about.

Creating child-classes of DukeException is more descriptive and
helps with organizing different exceptions.

ruiqi7 committed:
Use streams to find tasks with matching keywords

ruiqi7 committed:
Merge pull request #1 from ruiqi7/branch-A-Assertions

Add assertions to Parser and Storage classes

ruiqi7 committed:
Merge branch 'master' of https://github.com/ruiqi7/ip

ruiqi7 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/parser/Parser.java

ruiqi7 committed:
Merge pull request #2 from ruiqi7/branch-A-CodeQuality

Improve code quality

ruiqi7 committed:
Merge branch 'master' into branch-A-Streams

ruiqi7 committed:
Merge pull request #3 from ruiqi7/branch-A-Streams

Use streams to find tasks with matching keywords

ruiqi7 committed:
Merge branch 'master' of https://github.com/ruiqi7/ip

ruiqi7 committed:
Set up CI

ruiqi7 committed:
Add undo command

ruiqi7 committed:
Merge branch 'branch-C-Undo'

ruiqi7 committed:
Add OnCommand for users to find tasks on a specific date

ruiqi7 committed:
Improve GUI

* set background color
* change profile pictures and make them rounded
* customize chat bubbles
* add prompt text on text field

ruiqi7 committed:
Further improve GUI

* enable content resize when window resizes
* allow scrollpane to be pannable and scrollable
* highlight error messages with a different color

ruiqi7 committed:
Make minor modifications to UI messages

ruiqi7 committed:
Add more error handling

* prevent user from typing in an empty input
* trim white spaces from user input
* add more descriptive storage error messages

ruiqi7 committed:
Add more personality

ruiqi7 committed:
Update README

ruiqi7 committed:
Refactor code

* remove Messages class and shift the string constants into the classes they are related to
* shift exception classes into the packages they are related to
* add DateParser class to handle date formatting

ruiqi7 committed:
Add more JUnit tests

ruiqi7 committed:
Fix UndoCommand bug

ruiqi7 committed:
Add product screenshot

ruiqi7 committed:
Update README
